File: CS/SaveCheckStateTreeListNode(E3376)/SaveCheckStateTreeListNode/SaveCheckHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using DevExpress.XtraTreeList;
using System.Data;
using System.Windows.Forms;
using DevExpress.XtraTreeList.Columns;

namespace SaveCheckStateTreeListNode
{
    class SaveCheckHelper
    {
        readonly TreeList _treeList;
        readonly DataTable _source;
        readonly int _indexCheckColumn;

        public SaveCheckHelper(TreeList treeList, string field)
        {
            _treeList = treeList;
            TreeListColumn column = _treeList.Columns.ColumnByFieldName(field);
            if (column == null)
                throw new Exception(String.Format("TreeList doesn't contain the {0} field", field));
            _source = _treeList.DataSource as DataTable;
            _indexCheckColumn = column.ColumnHandle;
            column.Visible = false;
            SubscribeEvents();
        }

        private void SubscribeEvents()
        {
            _treeList.AfterCheckNode += _treeList_AfterCheckNode;
            _treeList.NodesReloaded += _treeList_NodesReloaded;
        }

        void _treeList_NodesReloaded(object sender, EventArgs e)
        {
            if (_treeList.DataSource == null)
                return;
            for (int i = 0; i < _source.Rows.Count; i++)
            {
                _treeList.FindNodeByID(i).CheckState = (CheckState)_source.Rows[i][_indexCheckColumn];
            }
        }

        void _treeList_AfterCheckNode(object sender, NodeEventArgs e)
        {
            if (_treeList.DataSource == null)
                return;
            _treeList.BeginUpdate();
            _source.Rows[e.Node.Id][_indexCheckColumn] = _treeList.FindNodeByID(e.Node.Id).CheckState;
            _treeList.EndUpdate();
        }
    }
}

File: VB/SaveCheckStateTreeListNode(E3376)/SaveCheckStateTreeListNode/SaveCheckHelper.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports DevExpress.XtraTreeList
Imports System.Data
Imports System.Windows.Forms
Imports DevExpress.XtraTreeList.Columns

Namespace SaveCheckStateTreeListNode
    Friend Class SaveCheckHelper
        Private ReadOnly _treeList As TreeList
        Private ReadOnly _source As DataTable
        Private ReadOnly _indexCheckColumn As Integer

        Public Sub New(ByVal treeList As TreeList, ByVal field As String)
            _treeList = treeList
            Dim column As TreeListColumn = _treeList.Columns.ColumnByFieldName(field)
            If column Is Nothing Then
                Throw New Exception(String.Format("TreeList doesn't contain the {0} field", field))
            End If
            _source = TryCast(_treeList.DataSource, DataTable)
            _indexCheckColumn = column.ColumnHandle
            column.Visible = False
            SubscribeEvents()
        End Sub

        Private Sub SubscribeEvents()
            AddHandler _treeList.AfterCheckNode, AddressOf _treeList_AfterCheckNode
            AddHandler _treeList.NodesReloaded, AddressOf _treeList_NodesReloaded
        End Sub

        Private Sub _treeList_NodesReloaded(ByVal sender As Object, ByVal e As EventArgs)
            If _treeList.DataSource Is Nothing Then
                Return
            End If
            For i As Integer = 0 To _source.Rows.Count - 1
                _treeList.FindNodeByID(i).CheckState = CType(_source.Rows(i)(_indexCheckColumn), CheckState)
            Next i
        End Sub

        Private Sub _treeList_AfterCheckNode(ByVal sender As Object, ByVal e As NodeEventArgs)
            If _treeList.DataSource Is Nothing Then
                Return
            End If
            _treeList.BeginUpdate()
            _source.Rows(e.Node.Id)(_indexCheckColumn) = _treeList.FindNodeByID(e.Node.Id).CheckState
            _treeList.EndUpdate()
        End Sub
    End Class
End Namespace
